crates/block-producer/src/block_builder/mod.rs

```diff
@@ -143,3 +145,33 @@ where
         Ok(())
     }
 }
+
+struct BlockProducer {
+    pub mempool: Arc<Mutex<Mempool>>,
+    pub block_interval: tokio::time::Duration,
+}
+
+impl BlockProducer {
+    pub async fn run(self) {
+        let mut interval = tokio::time::interval(self.block_interval);
+        interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Delay);
+
+        loop {
+            interval.tick().await;
+
+            let (block_number, batches) = self.mempool.lock().await.select_block();
+
+            let result = self.build_and_commit_block(batches).await;
+            let mut mempool = self.mempool.lock().await;
+
+            match result {
+                Ok(_) => mempool.block_committed(block_number),
+                Err(_) => mempool.block_failed(block_number),
+            }
+        }
+    }
+
+    async fn build_and_commit_block(&self, batches: BTreeSet<BatchJobId>) -> Result<(), ()> {
+        todo!("Aggregate, prove and commit block");
```

Comment:
This will contain building the block and saving it to the store, right?

For block building, since we don't actually prove anything yet, I am thinking we could put in an artificial delay of 3 - 5 seconds to get the conditions close to the real thing.

Reply:
Correct. I was also considering injecting failures randomly as mentioned [here](https://github.com/0xPolygonMiden/miden-node/pull/502#discussion_r1789644365).